Added the ability to extend lz.n with custom handlers. (#17)
nvim-neorocks/lz.n: README.md

  ├── init.lua
```

### Custom Handlers

You may register your own handlers to lazy-load plugins via
other triggers not already covered by the plugin spec.

You should register all handlers before calling `require('lz.n').load`,
because they will not be retroactively applied to
the `load` calls that occur before they are registered.

The `register_handler` function returns a boolean that indicates success.

`require("lz.n").register_handler(handler: lz.n.Handler): boolean`

#### lz.n.Handler

<!-- markdownlint-disable MD013 -->
| Property | Type | Description |
|----------|------|-------------|

> | spec_field | `string` | the new spec field defined by the handler |

Reviewer: ```suggestion
| spec_field | `string` | the `lz.n.PluginSpec` field defined by the handler |
```

  ├── init.lua
```

### Custom Handlers

You may register your own handlers to lazy-load plugins via
other triggers not already covered by the plugin spec.

> When using custom handlers, you must put `lazy = true` in your plugin spec.
> Otherwise, unless it has one of the other handler types already enabled,
> it will be eagerly loaded.
> 
> However, upon registering the handler,
> you may add names of spec items to check for automatically,
> thus removing this requirement.

Reviewer: suggestion: this doesn't convey that the spec fields can also be used to configure the handler.

```suggestion
Upon registering a handler,
you may add names of fields that extend the `lz.n.PluginSpec`,
which can be used to configure it.
If your handler does not register with a spec field, 
users must put `lazy = true` in the plugin spec.
Otherwise, unless it has one of the other handler types enabled, it will be eagerly loaded.
```

Author: Hmmmmmm

Should the handler itself register these fields, or should it be on the user when they call register_handler?

Author: because right now its on the user

Reviewer: I think it's fine to have it passed in to `register_handler`.

Author: I found a better solution. The author of the handler now adds them.

> You should register all handlers before calling load,

Reviewer: ```suggestion
You should register all handlers before calling `load`,
```

because they will not be retroactively applied to

> the load calls that occur before they are registered.

Reviewer: ```suggestion
the `load` calls that occur before they are registered.
```

`require("lz.n").register_handler(handler: lz.n.Handler)`

`require("lz.n").register_handler(handler: lz.n.Handler, spec_items: string|string[]|nil)`

#### lz.n.Handler

| Property | Type | Description |
|----------|------|-------------|
| type | `string` | the name of the handler |
| add | `fun(plugin: lz.n.Plugin)` | adds a plugin to the handler |
| del | `fun(plugin: lz.n.Plugin)?` | removes a plugin from the handler |

When writing custom handlers,
you can load the plugin and run the hooks from
the spec with the following function.

The function accepts plugin names or plugin specs.
It will call the handler's del function (if it exists) after the `before` hooks,
and before `load` of the plugin's spec.

```lua
  ---@type fun(plugins: string | lz.n.Plugin | string[] | lz.n.Plugin[])

>   require('lz.n').load_plugins

Reviewer: suggestion: How about `trigger_load` as a name?
I think that clarifies more clearly that this causes a pending plugin to be loaded.

> When writing custom handlers,
> you can load the plugin and run the hooks from
> the spec with the following function.
> 
> The function accepts plugin names or plugin specs.
> It will call the handler's del function (if it exists) after the `before` hooks,
> and before `load` of the plugin's spec.
> 
> ```lua
>   ---@type fun(plugins: string | lz.n.Plugin | string[] | lz.n.Plugin[])
>   require('lz.n').load_plugins
> ```

Reviewer: ```suggestion
When writing custom handlers,
you can load the plugin and run the hooks from
the spec with the following function:

```lua
  ---@type fun(plugins: string | lz.n.Plugin | string[] | lz.n.Plugin[])
  require('lz.n').load_plugins
```

The function accepts plugin names or plugin specs.
It will call the handler's `del` function (if it exists) after the `before` hooks,
and before `load` of the plugin's spec.
```

  ├── init.lua
```

### Custom Handlers

You may register your own handlers to lazy-load plugins via other triggers not already covered by the plugin spec.

When you want to add a new value to the set, it should be added to the extras table in the plugin spec.

`require("lz.n").register_handler(handler: lz.n.HandlerSpec)`

| Property | Type | Description |
|----------|------|-------------|
| type | `string` | the name of the handler |

> | pending | `table<string, table<string, string>>` | `{ key = { plugin.name = string }` the list of pending actions for the handler |

Author: Is my description of this type correct?
